### PR TITLE
Allow hyphens in cookie domain

### LINKED ISSUE #13: Cookie Domain Should Allow Hyphens
`func checkDomain` currently returns an "invalid character" error if any characters in the domain are not an ANSI letter, digit, or period. Hyphens should also be allowed in certain positions for domains such as `www.foo-bar.com`.

The func comment references [RFC 1034 Sec 3.5](https://tools.ietf.org/html/rfc1034#section-3.5). According to the RFC each label - the name portion between each period - (1) must begin with a letter, (2) must end with a letter or digit if at least 2 characters long, and (3) may contain a letter, digit, or hyphen for intermediate characters when at least 3 characters long.

I'm working on a patch right now for my own use and will submit a PR for consideration once complete. I don't see a `CONTRIBUTING.md` file. Are there any specific requirements besides updating the `TestCheckDomain` test?

### PR DESCRIPTION
**Pull request by blitzrk**
Resolves #13 

Follow RFC 1034 Section 3.5 by allowing hyphens in the middle of domain labels (period-separated parts).

### CODE REVIEW

File: securecookie.go [hunk 1]
```diff
@@ -205,27 +206,32 @@ func checkPath(path string) error {
...
+	case isDigit(name[0]):
+		return errors.New("cookie domain: starts with digit")
+	case name[len(name)-1] == '.':
+		return errors.New("cookie domain: ends with '.'")
+	}
+	for i, c := range name {
+		if utf8.RuneLen(c) != 1 {
+			return fmt.Errorf("cookie domain: invalid character '%c' at offset %d", c, i)
+		}
 		if c == '.' {
```
Comment (chmike) [inline securecookie.go]: I would use a byte variable after this for clarity:
ˋb := byte(c)ˋ

File: securecookie.go [hunk 2]
```diff
@@ -205,27 +206,32 @@ func checkPath(path string) error {
...
-			if i == 0 {
-				return errors.New("cookie domain: starts with '.'")
-			}
```
Comment (chmike) [inline securecookie.go]: I owuld then rename ˋpc` in ˋpb` for previous byte.

File: securecookie.go [hunk 3]
```diff
@@ -205,27 +206,32 @@ func checkPath(path string) error {
...
 			if pc := name[i-1]; pc == '-' || pc == '.' {
 				return fmt.Errorf("cookie domain: invalid character '%c' at offset %d", pc, i-1)
 			}
-			if i == len(name)-1 {
-				return errors.New("cookie domain: ends with '.'")
-			}
-			if nc := name[i+1]; nc == '-' {
+			if nc := name[i+1]; nc == '-' || isDigit(nc) {
 				return fmt.Errorf("cookie domain: invalid character '%c' at offset %d", nc, i+1)
 			}
```
Comment (chmike) [inline securecookie.go]: Ideally, ˋnc` should be the next rune and not a byte.

File: securecookie.go (hunk 1 above)
Comment (chmike) [reply securecookie.go]: Or use a rune everywhere which would be simpler. I don't think it would harm performance.

File: securecookie_test.go [hunk 4]
```diff
@@ -55,9 +55,21 @@ func TestCheckPath(t *testing.T) {
...
+	if err := checkDomain("example1.com"); err != nil {
+		t.Errorf("unexpected error: %s", err)
+	}
+	if err := checkDomain("foo-bar.com"); err != nil {
+		t.Errorf("unexpected error: %s", err)
+	}
+	if err := checkDomain("www1.foo-bar.com"); err != nil {
+		t.Errorf("unexpected error: %s", err)
+	}
 	if err := checkDomain(strings.Repeat("a", 300)); err == nil {
```
Comment (chmike) [inline securecookie_test.go]: We need to add a test with uppercase letters.

File: securecookie.go [hunk 5]
```diff
@@ -205,27 +206,32 @@ func checkPath(path string) error {
 // See https://tools.ietf.org/html/rfc1034#section-3.5 and
 // https://tools.ietf.org/html/rfc1123#section-2.
 func checkDomain(name string) error {
-	if len(name) > 255 {
+	switch {
+	case len(name) == 0:
+		return nil // an empty domain name will result in a cookie without a domain restriction
+	case len(name) > 255:
```
Comment (chmike) [inline securecookie.go]: A label is also restricted to be at most 63 byte long. This is currently unchecked.

**Comment by blitzrk** [comment]
There were some edge cases - covered by the tests I added - which weren't handled by the pseudocode in issue #13 such as a domain name ending in a hyphen. I handled these cases, but the code is no longer as concise. Any suggested improvements in this regard or in regards to error messages are very welcome.

**Comment by chmike** [comment]
It's looking much better. 

Since `isLetterOrDigit` Is used only in only in `isValidNameChar`, we can replace the call in this function with it's body. We should sort individual tests by decreasing validity frequency. Thus I suppose the test for lowercase letter should comme first, test for digit next, and then test for uppercase letter next.

Here is modified version of your code which I think is simpler. I currently only have an iPad and can't compile and run anything.

The rationale is to copy and adapt the switch statement that checks the domain label validity. 

```Go
// checkDomain returns an error if the domain name is not valid
// See https://tools.ietf.org/html/rfc1034#section-3.5 and
// https://tools.ietf.org/html/rfc1123#section-2.
func checkDomain(name string) error {
	switch {
	case len(name) == 0:
		return nil // an empty domain name will result in a cookie without a domain restriction
	case len(name) > 255:
		return fmt.Errorf("cookie domain: name length is %d, can't exceed 255", len(name))
	}
	var l int
	for i := 0; i < len(name); i++ {
		b := name[i]
		if b == '.' {
                        // check subdomain label validity
			switch {
			case i == l:
				return fmt.Errorf("cookie domain: invalid character '%c' at offset %d: label cannot begin with period", b, i)
                        case i-l > 53:
                                return fmt.Errorf("cookie domain: byte length of label '%s' is %d, can't exceed 63", name[l:i], i-l)
			case name[l] == '-':
				return fmt.Errorf("cookie domain: label '%s' at offset %d begins with a hyphen", name[l:i], l)
			case name[i-1] == '-':
				return fmt.Errorf("cookie domain: label '%s' at offset %d ends with a hyphen", name[l:i], l)
			}
			l = i + 1
			continue
		}
		// test label character validity, note: tests are ordered by decreasing validity frequency
		if !(b >= 'a' && b <= 'z' || b >= '0' && b <= '9' || b == '-' || b >= 'A' && b <= 'Z') {
			// show the printable unicode character starting at byte offset i
			c, _ := utf8.DecodeRuneInString(name[i:])
			if c == utf8.RuneError {
				return fmt.Errorf("cookie domain: invalid rune at offset %d", i)
			}
			return fmt.Errorf("cookie domain: invalid character '%c' at offset %d", c, i)
		}
	}
        // check top level domain validity
        switch {
        case l == len(name):
                return fmt.Errorf("cookie domain: missing top level domain, domain can't end with a period")
        case len(name) - l > 53:
                return fmt.Errorf("cookie domain: byte length of top level domain '%s' is %d, can't exceed 63", name[l], len(name)-l)
        case name[l] == '-':
                return fmt.Errorf("cookie domain: top level domain '%s' at offset %d begins with a hyphen", name[l:], l)
        case name[len(name)-1] == '-':
                return fmt.Errorf("cookie domain: top level domain '%s' at offset %d ends with a hyphen", name[l:], l)
        case name[l] >= '0' && name[l] <= '9' {
		return fmt.Errorf("cookie domain: top level domain '%s' at offset %d begins with a digit", name[l:], l)
	}
	return nil
}
```

**Comment by blitzrk** [comment]
It passes all the new tests, so it works for me! Updating the PR now.

**Comment by chmike** [comment]
There is a small mistake in my code in the second case of the top level domain check (after the `for` loop). The second argument of the fmt.Errorf should be `name[l:]` and not `name[l]`.

**Comment by blitzrk** [comment]
I caught that and already fixed it in my latest version. 😄 I did, however, mess up the order of the inlined `isLetterOrDigit`, so I'll fix that. Anything else?

**Comment by chmike** [comment]
Excellent work. It fixed an unexpected number of problems in my domain name checking function.
Thank you for your significant contribution.
